File: ALD.LibFiscalCode.Persistence/Models/Person.cs
using System;
using ALD.LibFiscalCode.Persistence.Enums;
using ALD.LibFiscalCode.Persistence.EqualityComparers;
using ALD.LibFiscalCode.Persistence.Events;

namespace ALD.LibFiscalCode.Persistence.Models
{
    public class Person : AbstractNotifyPropertyChanged
    {
        private readonly PersonEqualityComparer equalityComparer;
        private DateTime _dateOfBirth;

        private string _name;
        private string _surname;

        private Place placeOfBirth;

        public Person(DateTime defaultDateOfBirth) : this()
        {
            DateOfBirth = defaultDateOfBirth;
        }

        public Person()
        {
            Name = "";
            Surname = "";

            DateOfBirth = DateTime.Today;
            Gender = Gender.Unspecified;
            PlaceOfBirth = null;
            equalityComparer = new PersonEqualityComparer();
        }

        public FiscalCodeEntity FiscalCode { get; set; }

        public int Id { get; set; }

        public string Name
        {
            get => _name;
            set
            {
                _name = value;
                OnPropertyChanged(nameof(Name));
            }
        }

        public string Surname
        {
            get => _surname;
            set
            {
                _surname = value;
                OnPropertyChanged(nameof(Surname));
            }
        }

        public DateTime DateOfBirth
        {
            get => _dateOfBirth;
            set
            {
                _dateOfBirth = value;
                OnPropertyChanged(nameof(DateOfBirth));
                OnPropertyChanged(nameof(PrettyDate));
            }
        }

        public Place PlaceOfBirth
        {
            get => placeOfBirth;
            set
            {
                placeOfBirth = value;
                OnPropertyChanged(nameof(PlaceOfBirth));
            }
        }

        public Gender Gender { get; set; }

        public string PrettyDate => DateOfBirth.Date.ToShortDateString();

        public bool Equals(Person other)
        {
            return equalityComparer.Equals(this, other);
        }

        public PersonEqualityComparer GetEqualityComparer()
        {
            return equalityComparer;
        }
    }
}

File: CodFiscale.Online/src/models/Person.ts
import {Gender} from "@/models/Gender";
import {Place} from "@/models/Place";

export class Person {
    private name: string | null;
    private surname: string | null;
    private gender: Gender;
    private birthDate: Date | null;
    private birthPlaceId: number | null;
    get BirthPlaceId() {
        return this.birthPlaceId;
    }

    set BirthPlaceId(value) {
        this.birthPlaceId = value;
    }
    get BirthDate(){
        return this.birthDate;
    }

    set BirthDate(value) {
        this.birthDate = value;
    }
    get Gender() {
        return this.gender;
    }
    set Gender(value: Gender | string) {
        // Smelly
       if (typeof(value) == "string") {
           const initial = value.substr(0, 1).toUpperCase();
           if (initial === 'F') {
               this.gender = Gender.Female;
           }
           else if (initial === 'M') {
               this.gender = Gender.Male;
           }
           else {
               this.gender = Gender.Unspecified;
           }
       }
       else {
           this.gender = value;
       }

    }
    get Surname() {
        return this.surname;
    }

    set Surname(value) {
        this.surname = value;
    }

    get Name() {
        return this.name;
    }

    set Name(value) {
        this.name = value;
    }

    constructor() {
        this.name = "";
        this.surname = "";
        this.gender = Gender.Unspecified;
        this.birthDate = null;
        this.birthPlaceId = 0;
    }
}
